src/crypto/store/indexeddb-crypto-store.js

```diff
@@ -110,6 +111,238 @@ export default class IndexedDBCryptoStore {
             };
         });
     }
+
+    /**
+     * Look for an existing outgoing room key request, and if none is found,
+     * add a new one
+     *
+     * @param {module:crypto/store/base~OutgoingRoomKeyRequest} request
+     *
+     * @returns {Promise} resolves to
+     *    {@link module:crypto/store/base~OutgoingRoomKeyRequest}: either the
+     *    same instance as passed in, or the existing one.
+     */
+    getOrAddOutgoingRoomKeyRequest(request) {
+        const requestBody = request.requestBody;
+
+        return this.connect().then((db) => {
+            const deferred = q.defer();
+            const txn = db.transaction("outgoingRoomKeyRequests", "readwrite");
+            txn.onerror = deferred.reject;
+
+            // first see if we already have an entry for this request.
+            this._getOutgoingRoomKeyRequest(txn, requestBody, {
+                onFound: (existing) => {
+                    // this entry matches the request - return it.
+                    console.log(
+                        `already have key request outstanding for ` +
+                        `${requestBody.room_id} / ${requestBody.session_id}: ` +
+                        `not sending another`,
+                    );
+                    deferred.resolve(existing);
+                },
+                onNotFound: () => {
```

Comment:
API shape feels weird. Is there any reason why we don't invoke a single function and give a `null` parameter if it was not found?